src/gui/log/logmodel.h

```diff
@@ -0,0 +1,97 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2020  Prince Gupta <jagannatharjun11@gmail.com>
+ * Copyright (C) 2019  sledgehammer999 <hammered999@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QAbstractListModel>
+#include <deque>
+
+namespace Log
+{
+    struct Msg;
+    struct Peer;
+}
+
+class BaseLogModel : public QAbstractListModel
+{
+    Q_OBJECT
+    Q_DISABLE_COPY(BaseLogModel)
+
+public:
+    enum DataRole
+    {
+        TimeRole = Qt::UserRole + 1,
+        MessageRole
+    };
+    
+    using QAbstractListModel::QAbstractListModel;
+
+    int rowCount(const QModelIndex &parent = {}) const override;
+    int columnCount(const QModelIndex &parent = {}) const override;
+    QVariant data(const QModelIndex &index, int role = Qt::DisplayRole) const override;
+
+public slots:
+    void reset();
+
+protected:
+    struct Message
+    {
+        QVariant time;
+        QVariant message;
+        QVariant foregroundRole;
+        QVariant userRole;
+    };
+
+    void addNewMessage(Message message);
+
+private:
+    std::deque<Message> m_messages; // when logger is full, pop_front may take considerable time on vector like container
```

Comment:
This is not good enough, you should use circular buffer (same one from logger) as suggested before.

Reply:
boost::circular_buffer_optimesed causing crashing 
![image](https://user-images.githubusercontent.com/34717789/78901916-a62ef600-7a96-11ea-9bd2-b402b6252140.png)

can't seem to understand why, this happens at end i.e closing the main window, std::deque works fine.

Reply:
>can't seem to understand why

You should at least trace it to realize what code point couses the crash.

Reply:
I did trace it, it was here 
https://github.com/qbittorrent/qBittorrent/blob/43e5e242ff31da009c67e4365925e11604131980/src/gui/mainwindow.cpp#L792

I spent an hour on this, only after that I posted here.

Reply:
I suppose it's due to incorrect order of class destructions. First views should be destroyed and then models they use.

Reply:
Aren't you able to achieve more depth? Try setting breakpoints in your class destructors and see what happens, in what order they are removed, and how they interact with each other.

Reply:
hmm changing peer model and peer view initialization order seems to fixed it thanks.

but  I have another problem 
```
void BaseLogModel::addNewMessage(const BaseLogModel::Message& message)
{
    beginInsertRows(QModelIndex(), 0, 0);
    m_messages.push_back(message);
    endInsertRows();

    const int count = rowCount();
    if (count > MAX_LOG_MESSAGES) {
        const int lastMessage = count - 1;
        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
        endRemoveRows();
    }
}
```
after switching to circular buffer,
`if (count > MAX_LOG_MESSAGES) {...}` is not needed anymore right, since pushing on the filled circular buffer, will not change the size anymore and I also then don't need to `beginRemoveRows`, I can't fit this in my mind.

Reply:
I've pushed changes with buffer implementation, I tested with setting MAX_LOG_MESSAGES=20 and other small values.

Reply:
Actually the real reason of crash was that I left some std::move calls, sorry foe troubling you guys.